lib/jellyfish/room.ex

```diff
@@ -0,0 +1,186 @@
+
+    components =
+      engine_endpoints
+      |> Enum.filter(fn endpoint ->
+        Map.has_key?(state.components, endpoint.id)
+      end)
+      |> Map.new(&{&1.id, &1})
+
+    {:reply, %{id: state.id, peers: peers, components: components, config: state.config}, state}
+  end
+
+  def handle_call({:add_peer, peer_type}, _from, state) do
+    if Enum.count(state.peers) == state.config.max_peers do
+      {:reply, {:error, :reached_peers_limit}, state}
+    else
+      options = %{engine_pid: state.engine_pid, network_options: state.network_options}
+      peer = Peer.create_peer(peer_type, options)
+
+      state = put_in(state, [:peers, peer.id], peer)
+
+      :ok = Engine.add_endpoint(state.engine_pid, peer.engine_endpoint, endpoint_id: peer.id)
+
+      {:reply, peer, state}
+    end
+  end
+
+  def handle_call({:add_component, component_type, options}, _from, state) do
+    component =
+      Component.create_component(component_type, options, %{
+        engine_pid: state.engine_pid,
+        room_id: state.id
+      })
+
+    state = put_in(state, [:components, component.id], component)
+
+    :ok =
+      Engine.add_endpoint(state.engine_pid, component.engine_endpoint, endpoint_id: component.id)
+
+    {:reply, component, state}
+  end
+
+  def handle_call({:remove_component, component_id}, _from, state) do
+    {result, state} =
+      if Map.has_key?(state.components, component_id) do
+        {_elem, state} = pop_in(state, [:components, component_id])
+        {:ok, state}
+      else
+        {:error, state}
+      end
+
+    {:reply, result, state}
+  end
+
+  @spec get_state(room_pid :: pid()) :: t()
+  def get_state(room_pid) do
+    GenServer.call(room_pid, :state)
+  end
+
+  @spec add_peer(room_pid :: pid(), peer_type :: Peer.peer_type()) :: Peer.t() | {:error, any()}
+  def add_peer(room_pid, peer_type) do
+    GenServer.call(room_pid, {:add_peer, peer_type})
+  end
+
+  @spec add_component(
+          room_pid :: pid(),
+          component_type :: Component.component_type(),
+          options :: any()
+        ) :: Component.t()
+  def add_component(room_pid, component_type, options) do
+    GenServer.call(room_pid, {:add_component, component_type, options})
+  end
+
+  @spec remove_component(room_pid :: pid(), component_id :: String.t()) :: :ok | :error
+  def remove_component(room_pid, component_id) do
+    GenServer.call(room_pid, {:remove_component, component_id})
+  end
+
+  defp new(max_peers) do
+    id = UUID.uuid4()
+
+    rtc_engine_options = [
+      id: id
+    ]
+
+    {:ok, pid} = Engine.start(rtc_engine_options, [])
+    Engine.register(pid, self())
+    Process.monitor(pid)
+
+    turn_cert_file =
+      case Application.fetch_env(:jellyfish, :integrated_turn_cert_pkey) do
+        {:ok, val} -> val
+        :error -> nil
+      end
+
+    turn_mock_ip = Application.fetch_env!(:jellyfish, :integrated_turn_ip)
+
+    # turn_ip = if @mix_env == :prod, do: {0, 0, 0, 0}, else: turn_mock_ip
```

Comment:
I am not sure about it. Because we use this line in Videoroom, but it doesn't make much sense to me, I decided to leave it until we create a prototype of jellyfish.

Reply:
@mickel8 maybe (surely, right?) you know something about this?